EVMJIT: Use EVM-C callback functions

chfast opened the pull request:
EVMJIT PR: https://github.com/ethereum/evmjit/pull/67

chfast committed:
Implement evm_query() function for EVM-C

Implement evm_query() function and expect it to be used by EVMJIT in case of accesing balance and blockhash. The prevous functions for balance and blockash can be removed. Update EVMJIT submodule to the state that reflect that changes.

chfast committed:
Expect EVMJIT to use evm_query() for extcode and sload

chriseth commented on libevm/JitVM.cpp:
```diff
@@ -14,11 +13,135 @@ namespace dev
+namespace
+{
+
+static_assert(sizeof(Address) == sizeof(evm_hash160),
```
I kind of remember that static_assert is not available on all our platforms. That might have changed if we move to a newer msvc version.

chfast replied:
I pretty sure static_assert is supported by VS 2013. However, I've seen some bugs there. Nevertheless, I hope we will be able to test it in the CIs. In case of trouble, we can disable that on Windows.

chriseth commented on libevm/JitVM.cpp:
```diff
@@ -14,11 +13,135 @@ namespace dev
+static_assert(alignof(Address) == alignof(evm_hash160),
+              "Address types alignment mismatch");
+
+evm_hash160 toEvmC(Address _addr)
```
Wouldn't it be better to pass by const ref here (also below)?

chfast replied:
I really hope the compiler will inline that, then it does not matter. But good point. I planned to check it somehow.

chriseth commented on libevm/JitVM.cpp:
```diff
@@ -14,11 +13,135 @@ namespace dev
+evm_uint256 toEvmC(u256 _n)
+{
+	evm_uint256 m;
+	m.words[0] = static_cast<uint64_t>(_n);
```
Better to use `_n.convertTo<uint64_t>()`?

chfast replied:
Is it something different? If not I would prefer to keep generic expressions. How about exceptions here?

chriseth replied:
Ok, let's keep it like that.

chriseth commented on libevm/JitVM.cpp:
```diff
@@ -14,11 +13,135 @@ namespace dev
+	return u;
+}
+
+evm_variant evm_query(evm_env* _opaqueEnv, evm_query_key _key,
```
If you don't mind:

```
evm_variant evm_query(
    evm_env* _opaqueEnv,
    evm_query_key _key,
    evm_variant _arg
) noexcept
```

chfast replied:
I changed that but I think we should give up formatting that cannot be done by any tool. I tried to force clang-format to style like that, no success so far :)

chriseth commented on libevm/JitVM.cpp:
```diff
@@ -14,11 +13,135 @@ namespace dev
+		{
+			auto addr = fromEvmC(_arg.address);
+			auto &code = env.codeAt(addr);
+			v.bytes.bytes = reinterpret_cast<char const*>(code.data());
```
Is the lifetime of code long enough? I think it might be invalidated as soon as we do a recursive create, because that might cause a re-allocation of the m_cache map.

chfast replied:
True, looks scary. It used to work that way before https://github.com/ethereum/libethereum/blob/develop/libevm/JitEnv.cpp#L106-L112.

I think evmjit requires that to be valid only for a moment because the opcode is EXTCODECOPY and as the name suggests it copies the memory.

But this is good point and this case has to be at least be documented in the EVM-C.

chfast commented:
Thanks very much for reviewing that. It is still work in progress. But that helped me identified some issues in the EVM-C API. I will also need to introduces more changes to the EVM-C.

I planned to merge this when all the callbacks are replaced with the ones specified in the EVM-C. I hope it will no be very big then. But we can discuss this approach.

chfast committed:
Expect EVMJIT to use evm_query() for accessing addresses

chfast committed:
Expect EVMJIT to use evm_query() for accessing tx/block info

chfast committed:
Implement evm_update() callback function

Expect EVMJIT to be using evm_update() function for updating a contract storage instead of env_sstore() one.

chfast committed:
Expect EVMJIT registering selfdestruction by evm_update() function

chfast committed:
Add EVM_LOG support in evm_update(), drop evn_log()

chfast committed:
Expect EVMJIT to use evm_call() for calls

chfast committed:
Change evm_call() signature

chfast committed:
evm_call: use EVM_EXCEPTION as the failure indicator

chfast committed:
Expect EVMJIT to use evm_call() for CREATE

chfast committed:
JitVM: mark helper function with inline

chfast committed:
Reformat code

gcolvin commented:
@chfast, I looked over the changes, especially the additions to JitVM.cpp.  They look good to me.

chfast commented:
@chriseth would you like to take a look again (not pushing)?

When is the best time to merge it?

chfast committed:
Merge remote-tracking branch 'upstream/develop' into evmc

chfast committed:
Do not use evm_bytes_view type in EVM-C

bobsummerwill commented:
We can merge whenever we are comfortable with the changes.

I don't think we need to hold-off until after 1.3.1, for example, @chfast.

It's trivial for me to keep my cpp-ethereum in sync with webthree-umbrella changes as they come in.    Don't feel that we need to stop working :-)

chfast commented:
Not merging that is not blocking me. However, we don't any CI with evmjit enabled.

bobsummerwill commented:
Right - because of the build times for LLVM.

chriseth commented:
If this 'works' (choose your definition), it can be merged.
